### PR TITLE
Secrets Enhancements

### PR DESCRIPTION
**Pull request by hemant0308**
* Feature to tag multiple projects to secret
* Migration task to update hash algorithm

### CODE REVIEW

File: server/db/src/main/resources/com/walmartlabs/concord/server/db/v1.96.0.xml [hunk 1]
```diff
@@ -47,4 +47,40 @@
             </column>
         </addColumn>
     </changeSet>
+
+    <changeSet id="1960012" author="g0h04k0walmart.com">
+        <createTable tableName="SECRET_PROJECTS">
```
Comment (ibodrov) [inline server/db/src/main/resources/com/walmartlabs/concord/server/db/v1.96.0.xml:52]: Maybe `PROJECT_SECRETS`? Otherwise it sounds like projects themselves are secret :-)

File: server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretEntry.java [hunk 2]
```diff
@@ -37,7 +37,7 @@
 @JsonInclude(Include.NON_NULL)
 public class SecretEntry implements Serializable {
 
-    private static final long serialVersionUID = 1L;
+    private static final long serialVersionUID = 2L;
```
Comment (ibodrov) [inline server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretEntry.java:40]: Why is this necessary?

File: console2/src/api/org/secret/index.ts [hunk 3]
```diff
@@ -184,19 +184,20 @@ export const updateSecretVisibility = (
 export const updateSecretProject = (
     orgName: ConcordKey,
     secretName: ConcordKey,
-    projectName: ConcordKey
+    projects: ProjectEntry[]
```
Comment (ibodrov) [inline console2/src/api/org/secret/index.ts:187]: Don't pass the whole ProjectEntry if the function uses only the IDs.

**Review by ibodrov — COMMENTED** [review]
This still feels like several PRs in one - e.g the migration task can be a separate PR. To avoid mixing maintenance PRs with feature PRs.

A few notes so far:

File: server/db/src/main/resources/com/walmartlabs/concord/server/db/v1.96.0.xml (hunk 1 above)
Comment (hemant0308) [reply server/db/src/main/resources/com/walmartlabs/concord/server/db/v1.96.0.xml]: Agree. :) Renamed to PROJECT_SECRETS .

**Comment by hemant0308** [comment]
> This still feels like several PRs in one - e.g the migration task can be a separate PR. To avoid mixing maintenance PRs with feature PRs.
Moved the migration task to another PR https://github.com/walmartlabs/concord/pull/691

**Comment by brig** [comment]
As for me, adding a new field to the API entity is not a breaking change

**Comment by hemant0308** [comment]
@walmartlabs/concord Can someone review it. i am planning to send this change on next release.

File: console2/src/components/organisms/SecretProjectActivity/index.tsx [hunk 4]
```diff
@@ -18,93 +18,118 @@
...
-                        primary={true}
-                        negative={true}
-                        content="Update"
-                        disabled={!dirty}
-                        onClick={() => setShowConfirm(true)}
-                    />
                 </Form.Group>
-
+                ):null}
+                <Form.Group>
```
Comment (benbroadaway) [inline console2/src/components/organisms/SecretProjectActivity/index.tsx:94]: There's a UI bug where the views are not updated after the projects are updated. The immediate view _is_ updated, but parent and neighbors are not.


https://user-images.githubusercontent.com/4554569/225025941-691aeb09-98d2-4c53-94be-66f56e11c5b7.mov



A page refresh sorts things out, but there's a case where things can be confusing and re-add unexpected projects if the user doesn't refresh. An example
1. remove an existing project from a secret
2. immdate view is updated to reflect correct projects (or none)
3. go to another tab, e.g. `Info` -- `Restricted to a project` element is stale  
4. Return to `Settings` tab -- now the Projects view is stale, showing the old project that was removed
5. Add another project -- now the Secret's Projects have been updated to include the previously removed project plus the new one




https://user-images.githubusercontent.com/4554569/225025156-b28150c2-bea7-4d57-ac39-7fbce28c7463.mov

Comment (hemant0308) [reply console2/src/components/organisms/SecretProjectActivity/index.tsx:110]: https://user-images.githubusercontent.com/13145963/225324572-a3a9daf4-857b-460d-8bba-40f2bb7b68a5.mov

Thanks Ben for catching this bug. resolved it.

File: console2/src/api/org/secret/index.ts [hunk 5]
```diff
@@ -184,19 +185,20 @@ export const updateSecretVisibility = (
 export const updateSecretProject = (
     orgName: ConcordKey,
     secretName: ConcordKey,
-    projectName: ConcordKey
+    projectIds: String[]
 ): Promise<GenericOperationResult> => {
+    const data = new FormData();
+    if(projectIds.length > 0) {
```
[the remark this answers is not in the document]
Comment (hemant0308) [reply console2/src/api/org/secret/index.ts]: Formatted only the changes files in this PR with prettify

File: server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretUpdateParams.java [hunk 6]
```diff
@@ -37,10 +38,7 @@ public interface SecretUpdateParams extends Serializable {
     String newOrgName();
 
     @Nullable
-    UUID newProjectId();
-
-    @Nullable
-    String newProjectName();
```
Comment (benbroadaway) [inline server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretUpdateParams.java:43]: These need to stick around for now. There isn't really a good reason someone (outside of our project) should be using it, but since it's `public` this is technically a breaking change.

Annotate with `@Deprecated` and we can remove them in a subsequent release (post-JDK17 lang level?)

**Review by benbroadaway — CHANGES_REQUESTED** [review]

File: server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretUpdateParams.java (hunk 6 above)
Comment (hemant0308) [reply server/impl/src/main/java/com/walmartlabs/concord/server/org/secret/SecretUpdateParams.java]: Added these attributes back with deprecated annotation.

File: it/common/src/main/java/com/walmartlabs/concord/it/common/ServerClient.java [hunk 7]
```diff
@@ -97,18 +97,55 @@ public SecretOperationResponse generateKeyPair(String orgName, String projectNam
...
+        }
 
+        if (projectIds != null && !projectIds.isEmpty()) {
+            m.put("projectIds", projectIds.stream().map(UUID::toString).collect(Collectors.joining(",")));
+        } else if (projectNames != null && !projectNames.isEmpty()) {
+            m.put("projects", String.join(",", projectNames));
+        }
         return postSecret(orgName, m);
     }
+    public SecretOperationResponse addPlainSecret(String orgName, String name, String projectName, boolean generatePassword, String storePassword, byte[] secret) throws ApiException {
```
[the remark this answers is not in the document]
Comment (hemant0308) [reply it/common/src/main/java/com/walmartlabs/concord/it/common/ServerClient.java]: Updated

**Review by benbroadaway — APPROVED** [review]

**Review by brig — APPROVED** [review]

**Review by mtkunkel — APPROVED** [review]
